core/trino-main/src/main/java/io/trino/operator/window/matcher/IrRowPatternToProgramRewriter.java

```diff
@@ -0,0 +1,317 @@
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.IntStream;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.collect.Collections2.orderedPermutations;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static java.lang.String.format;
+import static java.util.Objects.requireNonNull;
+
+public class IrRowPatternToProgramRewriter
+{
+    private IrRowPatternToProgramRewriter() {}
+
+    public static Program rewrite(IrRowPattern node, Map<IrLabel, Integer> labelMapping)
+    {
+        List<Instruction> instructions = new ArrayList<>();
+        new Rewriter(instructions, labelMapping).process(node);
+        instructions.add(new Done());
+        return new Program(instructions);
+    }
+
+    private static class Rewriter
+            extends IrRowPatternVisitor<Void, Void>
+    {
+        private final List<Instruction> instructions;
+        private final Map<IrLabel, Integer> labelMapping;
+
+        Rewriter(List<Instruction> instructions, Map<IrLabel, Integer> labelMapping)
+        {
+            this.instructions = requireNonNull(instructions, "instructions is null");
+            this.labelMapping = requireNonNull(labelMapping, "labelMapping is null");
+        }
+
+        @Override
+        protected Void visitIrRowPattern(IrRowPattern node, Void context)
+        {
+            throw new UnsupportedOperationException("unsupported node type: " + node.getClass().getName());
+        }
+
+        @Override
+        protected Void visitIrLabel(IrLabel node, Void context)
+        {
+            instructions.add(new MatchLabel(labelMapping.get(node)));
+            return null;
+        }
+
+        @Override
+        protected Void visitIrEmptyPattern(IrEmptyPattern node, Void context)
+        {
+            return null;
+        }
+
+        @Override
+        protected Void visitIrAnchorPattern(IrAnchorPattern node, Void context)
+        {
+            switch (node.getType()) {
+                case PARTITION_START:
+                    instructions.add(new MatchStart());
+                    return null;
+                case PARTITION_END:
+                    instructions.add(new MatchEnd());
+                    return null;
+                default:
+                    throw new IllegalStateException("unexpected anchor type: " + node.getType());
+            }
+        }
+
+        @Override
+        protected Void visitIrExcludedPattern(IrExcludedPattern node, Void context)
+        {
+            instructions.add(new Save());
+            process(node.getPattern());
+            instructions.add(new Save());
+            return null;
+        }
+
+        @Override
+        protected Void visitIrPatternAlternation(IrPatternAlternation node, Void context)
+        {
+            checkArgument(node.getPatterns().size() > 1, "invalid pattern: alternation with single element");
```

Comment:
This should be an invariant of the IrPatternAlternation node. See comment in that class